tests/unit_tests/config/test_forward_model.py

```diff
@@ -305,3 +307,304 @@ def test_that_eclipse100_require_version_field():
+
+    with pytest.raises(
+        ConfigValidationError,
+        match=".*This is a bad forward model step, dont use it.*",
+    ):
+        config.forward_model_data_to_json("id", 0, 0)
+
+
+def test_that_plugin_forward_model_raises_pre_experiment_validation_error_early(
+    tmp_path,
+):
+    (tmp_path / "test.ert").write_text(
+        """
+        NUM_REALIZATIONS  1
+        FORWARD_MODEL FM1(<arg1>=never,<arg2>=world,<arg3>=derpyderp)
+        FORWARD_MODEL FM2
+        """
+    )
+
+    class FM1(ForwardModelStepPlugin):
+        def __init__(self):
+            super().__init__(name="FM1", command=["the_executable.sh"])
+
+        def validate_pre_experiment(self, fm_step_json: ForwardModelStepJSON) -> None:
+            assert self.name == "FM1"
+            raise ChildProcessError("I don't think I wanna do hamster style anymore")
+
+    class FM2(ForwardModelStepPlugin):
+        def __init__(self):
+            super().__init__(
+                name="FM2",
+                command=["the_executable.sh"],
+            )
+
+        def validate_pre_experiment(self, fm_step_json: ForwardModelStepJSON) -> None:
+            assert self.name == "FM2"
+            raise PermissionError("well that's nice")
+
+    with pytest.raises(ConfigValidationError, match=".*hamster style.*that's nice.*"):
+        _ = ErtConfig.with_plugins(forward_model_step_classes=[FM1, FM2]).from_file(
+            tmp_path / "test.ert"
+        )
+
+
+def test_that_pre_run_substitution_forward_model_json_is_created_for_plugin_fms(
+    tmp_path,
+):
+    (tmp_path / "test.ert").write_text(
+        dedent(
+            """
+        NUM_REALIZATIONS  1
+
+        DEFINE <yo> dear
+        DEFINE <dawg> good
+        DEFINE <iherdulike> solonius
+        DEFINE <some_var> schmidt
+
+        FORWARD_MODEL FM1(<arg1>=<yo>,<arg2>=<dawg>,<arg3>=<iherdulike>)
+        """
+        )
+    )
+
+    class FM1(ForwardModelStepPlugin):
+        def __init__(self):
+            super().__init__(
+                name="FM1",
+                command=[
+                    "the_executable.sh",
+                    "sed",
+                    "-i",
+                    "<yo>",
+                    "-c",
+                    "<dawg>",
+                    "<iherdulike>",
+                    "<some_var>",
+                    "iter",
+                ],
+            )
+
+        def validate_pre_experiment(self, fm_step_json: ForwardModelStepJSON) -> None:
+            assert fm_step_json["argList"] == [
+                "sed",
+                "-i",
+                "dear",
+                "-c",
+                "good",
+                "solonius",
+                "schmidt",
+                "iter",
+            ]
+
+            # It is in the command, but not in the forward model(...) entry in the
+            # ert config. Thus it is not a "private" arg in that sense.
+            # Note2reviewer: We COULD include references to variables in the
```

Comment:
Maybe remove Note2reviewer and put that in github comments instead :D

Reply:
Ohh, I was trying to say that private_args will only include substituted args that are declared in the ert config, but not necessarily all the args in arglist.

So if arglist (in the fm step specification) is `[x,y,z]`, but the forward model invocation in the ert config is this `FORWARD_MODEL FM(<x>=2)`, then `self.private_args` will only contain `<x>`, and the concern was whether developers implementing a forward model would be interested in what `y` and `z` _would_ be substituted to, even though it isn't specified in the forward model step invocation in the ert config.

Reply:
So my question/concern was whether it was important to tweak the substitution logic to somehow make the substituted args from the entire arglist available, for validation purposes. If a developer implementing a fm step plugin says that `X` is in the arglist, and in the pre-experiment / pre-realization hook wants to know what it would be substituted with, regardless of whether it is invoked in the ert config fm step. I would assume this is never the case but am not completely sure.

Reply:
We looked into this a while ago and we are pretty much unable to check the entire arglist because it seems that some user scripts handle unsubstituted arguments on their own.